qa/python_models/bls/model.py

```diff
@@ -553,6 +555,86 @@ def test_timeout(self):
+
+        self.assertEqual(response_count, expected_output_cnt)
+
+        # Make sure the iterator is exhausted.
+        with self.assertRaises(StopIteration):
+            next(response_iterator)
+
+        return response_iterator
+
+    def test_response_iterator(self):
+        if self._is_decoupled:
+            # Test the response iterator for decoupled responses. The request
+            # has 4 decoupled responses.
+            response_num = 4
+            input0_np = np.array([response_num], dtype=np.int32)
+            input0 = pb_utils.Tensor('IN', input0_np)
+            infer_request = pb_utils.InferenceRequest(
+                model_name='square_int32',
+                inputs=[input0],
+                requested_output_names=['OUT'])
+            infer_responses = infer_request.exec(decoupled=True)
+
+            # case 1. Use Next() to get the next response first, then use
+            # for-loop to get the remaining responses.
+            infer_response = next(infer_responses)
+            self.assertFalse(infer_response.has_error())
+            output0 = pb_utils.get_output_tensor_by_name(infer_response, 'OUT')
+            self.assertIsNotNone(output0)
+            self.assertEqual(4, output0.as_numpy())
+            # The iterator now should only have 3 remaining responses.
+            infer_responses = self._test_response_iterator_square(
+                3, response_num, infer_responses)
+
+            # case 2. Call for-loop to get all the responses multiple times.
+            infer_responses = self._test_response_iterator_square(
+                4, response_num, infer_responses)
+            infer_responses = self._test_response_iterator_square(
+                4, response_num, infer_responses)
+            infer_responses = self._test_response_iterator_square(
+                4, response_num, infer_responses)
+
+            # case 3. Break from the iteration, then use Next() and for-loop to
+            # get the remaining responses.
+            response_count = 0
+            for infer_response in infer_responses:
+                self.assertFalse(infer_response.has_error())
+                output0 = pb_utils.get_output_tensor_by_name(infer_response,
+                                                             'OUT')
+                self.assertIsNotNone(output0)
+                self.assertEqual(4, output0.as_numpy())
+
+                response_count += 1
+                if response_count == 2:
+                    break
+
+            infer_response = next(infer_responses)
+            self.assertFalse(infer_response.has_error())
+            output0 = pb_utils.get_output_tensor_by_name(infer_response, 'OUT')
+            self.assertIsNotNone(output0)
+            self.assertEqual(4, output0.as_numpy())
+
+            # The iterator now should only have 1 remaining response.
+            infer_responses = self._test_response_iterator_square(
+                1, response_num, infer_responses)
```

Comment:
I think we also need to have a test for the case that the iterator is not fully used and we just delete it. 

```
delete inference_responses
```

I think it might be a bit difficult to confirm whether the responses have all been retrieved.